source/common/upstream/health_discovery_service.cc

```diff
@@ -54,12 +87,73 @@ void HdsDelegate::onReceiveInitialMetadata(Http::HeaderMapPtr&& metadata) {
   UNREFERENCED_PARAMETER(metadata);
 }
 
+void HdsDelegate::processMessage(
+    std::unique_ptr<envoy::service::discovery::v2::HealthCheckSpecifier>&& message) {
+  for (int i = 0; i < message->health_check_size(); i++) {
+    // Create HdsCluster config
+    ENVOY_LOG(debug, "Creating HdsCluster config");
+    envoy::api::v2::core::BindConfig bind_config;
+    envoy::api::v2::Cluster cluster_config_;
+    clusters_config_.push_back(cluster_config_);
+    cluster_config_.set_name("anna" + std::to_string(i));
+    cluster_config_.mutable_connect_timeout()->set_seconds(2);
+    cluster_config_.mutable_http2_protocol_options();
+    cluster_config_.mutable_per_connection_buffer_limit_bytes()->set_value(12345);
+
+    for (int j = 0; j < message->mutable_health_check(i)->endpoints_size(); j++) {
+      for (int k = 0; k < message->mutable_health_check(i)->mutable_endpoints(j)->endpoints_size();
+           k++) {
+        // Add endpoint to cluster
+        cluster_config_.add_hosts()->MergeFrom(*message->mutable_health_check(i)
+                                                    ->mutable_endpoints(j)
+                                                    ->mutable_endpoints(k)
+                                                    ->mutable_address());
+      }
+    }
+
+    for (int l = 0; l < message->mutable_health_check(i)->health_checks_size(); l++) {
+      // Add health checker to cluster
+      cluster_config_.add_health_checks()->MergeFrom(
+          *message->mutable_health_check(i)->mutable_health_checks(l));
+    }
+
+    ENVOY_LOG(debug, "New HdsCluster config {} ", cluster_config_.DebugString());
+
+    // Creating HdsCluster
+    cluster_.reset(new HdsCluster(runtime_, cluster_config_, bind_config, store_stats,
+                                  ssl_context_manager_, secret_manager_, false));
+
+    hds_clusters_.push_back(cluster_);
+    std::vector<Upstream::HealthCheckerSharedPtr> cluster_health_checkers;
```

Comment:
Yeah, I think there is also an issue here in considering the part of the message outside the scope of the specific cluster in the outer loop.